src/kernels/kernelProvider.node.ts

```diff
@@ -42,8 +43,18 @@ export class KernelProvider extends BaseKernelProvider {
     }
 
     public getOrCreate(uri: Uri, options: KernelOptions): IKernel {
-        const existingKernelInfo = this.getInternal(uri);
         const notebook = workspace.notebookDocuments.find((nb) => nb.uri.toString() === uri.toString());
+        // If we're connecting to a live local kernel, then Uri will point to the Uri of the original kernel.
+        if (options.metadata.kind === 'connectToLiveLocalKernel') {
```

Comment:
I feel like we've pushed the ownership of handling 'live' kernels up to the kernelProvider and the controller. With remote this is all handled by the kernel finder. 

Seems like we could do it the same way, we just need something in the raw kernel stuff to keep track of running kernels.

Reply:
Otherwise the controller and the kernelprovider have to special case live local kernels.

Reply:
ILocalKernelFinder could return running kernels too. In fact jupyter puts them in a known spot. We just need a way to store them for raw.

Reply:
> remote this is all handled by the kernel finder.

Don't think kernel finder should own lifetime of kernels, the name itself indicates it is responsible for finding.
KernelProvider IMHO is the right place as that starts kernels and tracks them and dispose them as well.

> ILocalKernelFinder could return running kernels too

Yes, that's possible.

> In fact jupyter puts them in a known spot. We just need a way to store them for raw.

Yes, i thought about that as well, but that requires bit more work, to track the connection json and exposing that and then starting a new kernel with the kernel information provided and that messes up a lot of other things. E.g. we have 2 processes but connected to the same underlying kernel and could complicate how we handle intellisense, widgets, etc. This way we just use the same IKernel and don't really start new sesions.
E.g. even in remote case, we do'nt really start new sessions, we just connect to the exact same controller from multiple notebooks & I'd prefer to do the same here as well.

Reply:
> ILocalKernelFinder could return running kernels too. In f

not going to do this, as thi will then need to know about live kernels & IKernel and the like.
I.e. this will need information about what s in KernelProvider.

Also mentioned earlier about issues with using that approach to start kernels. I think that complicates things a lot & could have more issues as well.

Reply:
I disagree. 

For non zmq, it works just like remote. No difference.
For zmq, we simply need a KernelProcessManager. It returns the list of running processes. Basically mimic how jupyter does it.

Reply:
In fact we could abstract out the session manager part to be ISessionManagerFactory and ISessionManager and the abstracted part only supports getting the list of running sessions.

Reply:
IKernelProvider would not be pushed down (as it isn't for IRemoteKernelFinder). Instead the raw code would have to expose a way to get the list of processes it started. Much like the jupyter session manager does.

Reply:
Actually it's more like the list of RawKernels/RawSessions (as we can't create another RawKernel to connect to the same process), but same idea as the JupyterSessions.

Reply:
> For zmq, we simply need a KernelProcessManager. It returns the list of runnin

Ok, I think I understand what yoo're suggesting, I thought you wanted me to spin up a new process and use the same old `kernel-*.json` connection file (which also works, but more complex).

Reply:
> as we can't create another RawKernel to connect to the same process

Actually we can, & that's what I thought you were suggesting all this time.
All we need to do is start a web socket connection for the same old kernel connection instead of starting a new process..

Reply:
Hey if that works, probably even better (so that dispose stuff doesn't get confused).